### ISSUE TITLE
[REMOVE] Privacy Badger

### ISSUE THREAD

**Issue by freddy-m** [tycrek/degoogle#426]
### Checklist

- [X] **I have read the guidelines in [CONTRIBUTING.md](../blob/master/CONTRIBUTING.md)**
- [X] Include my name in [CONTRIBUTORS.md](../blob/master/CONTRIBUTORS.md)
- [ ] I am affiliated with this alternative

### Alternative name

Privacy Badger

### Reason

It is rendered unnecessary by uBlock Origin - the fewer browser extensions the better.

**Comment by freddy-m** [comment]
Should add that it's very [fingerprintable](https://adtechmadness.wordpress.com/2020/03/27/detecting-privacy-badgers-canvas-fp-detection/), and also just not [needed](https://github.com/arkenfox/user.js/wiki/4.1-Extensions#small_orange_diamond-dont-bother).

**Comment by ghostwords** [comment]
**All** extensions that modify page state are fingerprintable. Canvas fingerprinting detection that the adtech madness article analyzes is disabled by default.

The arkenfox user.js page falsely claims that Privacy Badger no longer uses heuristics. Privacy Badger automatically learns to block invisible trackers regardless of whether you enable local learning. If you don't enable it locally, you periodically get  updated learnings from [Badger Sett](https://github.com/EFForg/badger-sett).

uBlock Origin is an effective privacy tool but it has its own distinct mission and feature set. Why push for a monoculture where all privacy tools are based on manually assembled lists?

**Comment by Thorin-Oakenpants** [comment]
> The arkenfox user.js page falsely claims that Privacy Badger no longer uses heuristics

Here's what the wiki entry says
> Note: Privacy Badger is easily [detected](https://adtechmadness.wordpress.com/2020/03/27/detecting-privacy-badgers-canvas-fp-detection/), and [no longer](https://www.eff.org/deeplinks/2020/10/privacy-badger-changing-protect-you-better) uses [hueristics](https://www.eff.org/privacybadger/faq#How-does-Privacy-Badger-work)

This is 100% correct. Read the links. Privacy Badger the extension (it's a wiki page about extensions) no longer uses heuristics. Rules discovered in any way shape or form and sent as updates periodically to the extension, is not heuristics being used by Privacy Badger the extension. Where the rules come from is immaterial to the statement on arkenfox

**Comment by ghostwords** [comment]
Anyone can re-enable local learning with one click on the options page. Imagine browsing the Web in a region poorly covered by manually assembled lists.

If you don't re-enable, you get periodic automatic updates generated by the same heuristics.

Privacy Badger may be detectable, but so is literally every other extension that modifies anything on a page. This includes 100% of content blockers.

The link you use to support the point that Privacy Badger is detectable no longer applies to Privacy Badger users who leave settings alone.

I cowrote the other article and the FAQ entry you suggest I read.

**Comment by Thorin-Oakenpants** [comment]
Sorry if this seems to be derailing the thread, but below are some more reasons why PB is redundant/obsolete for Firefox users

On the heuristics point, you're splitting hairs IMO. I am fully aware the option was left in - that's what the link is for, I did not wish to write "and no longer uses heuristics _**by default but you should never turn that on because the article says x y and z**_". Again, it does not matter how you derive the preset rules.

>Imagine browsing the Web in a region poorly covered by manually assembled lists.

Firefox users have State Partitioning
- [Network Partitioning](https://blog.mozilla.org/security/2021/01/26/supercookie-protections/) has been the default for all Firefox users since FF85
- [Total Cookie Protection](https://blog.mozilla.org/security/2021/02/23/total-cookie-protection/) which Mozilla is [rolling out to all users](https://bugzilla.mozilla.org/show_bug.cgi?id=1731713)
 - Arkenfox has used FPI for almost four years, but will move to the newer State Partitioning

Most (not all) cross-domain state tracking has been neutralized (and arkenfox sanitizes per session). In that context, Privacy Badger is obsolete, and even uBlock Origin is _almost_ a luxury: it has other benefits such as making the web more tolerable, improves perf, decreases the attack surface and some other features that help privacy. PB is obsolete these days as a form of enumerating badness

> [canvas] no longer applies to Privacy Badger users who leave settings alone

That leaves stateless. I am fully aware of how to fingerprint extensions ([prototypeLies](https://github.com/arkenfox/TZP/blob/master/js/prototypeLies.js)), as well as testing behavioral changes and responses to the document (the benefits of uBO far outweigh a potential but unlikely attack for filter list entropy). The point of that article, for me, was that it could be reversed - detecting tampering is trivial, as any known pixel test can prove. Besides, extensions lack the APIs to fully protect

Like state (state partitioning), for stateless I also prefer to work upstream with robust solutions built into the browser that actually address the problem. In that regard, PB is obsolete with RFP (which arkenfox uses)

I did not know about the change in canvas - was it disabled or removed? I do not see any setting for it. Can you elaborate? Except for "learn to block...", everything is enabled by default - not that I keep track of every extension, but I just checked a fresh install.

---

The entire wiki is being re-written, and PB falls (or fell) into an awkward cross between tracker blocking and fingerprinting. _**What would _you_ like it to say?_** - this is your chance :)

again, sorry if this is derailing the topic - I am more than happy to collapse my replies and discuss this at arkenfox

**Comment by ghostwords** [comment]
You seem to think Privacy Badger attempts to tamper with canvas fingerprints. It doesn't and it never did. What made you think that's the case?

>it has other benefits such as making the web more tolerable, improves perf, decreases the attack surface and some other features that help privacy

This also describes Privacy Badger. Please reread my previous replies in this thread.

I'll repeat this one point for the last time. Privacy Badger can and does learn to block trackers not present in manually assembled lists, regardless of whether the learning happened locally on your machine or not.

**Comment by Thorin-Oakenpants** [comment]
> You seem to think Privacy Badger attempts to tamper with canvas fingerprints

Sorry, my bad. It's been a while since I looked at it. The point was that it _was_ detectable - I even pointed to prototypeLies.

I am more than happy to remove that part when you clarify that it is not used and cannot be used in PB. You seemed to imply it only wasn't if users did not change settings

> This also describes Privacy Badger

Those are also redundant - PB is a _tracker_ blocker and state partitioning renders them obsolete. And I do not see them as anyway near comparable to uBlock's Origin's significantly larger _content blocking_ ability, and there are items I am referring to do not exist in PB. I'm not going to get into this off-topic tangent

> Privacy Badger can and does learn to block trackers

And I'll repeat this point one last time, it is irrelevant how you compile your predefined lists.

Instead of arguing, would you like the wiki to say? This is your chance

**Comment by ghostwords** [comment]
I don't think I can help you.

**Comment by Thorin-Oakenpants** [comment]
So you can't confirm (or correct) your own statements (emphasis mine)

> Privacy Badger's canvas fingerprinting detection that the adtech madness article analyzes is **_disabled by default_**

> The link you use to support the point that Privacy Badger is detectable _**no longer applies to Privacy Badger users who leave settings alone**_

Both implying that the feature is still there

It's not my extension, and I don't have the time nor inclination to test or follow hundreds of extensions when far better robust upstream solutions exist. You're the one that doesn't like the description on arkenfox, so I asked and offered to change it. All you need to do is confirm there is a setting (or not) for canvas - is it part of the local learning, which is the only feature not enabled by default? If not, then so be it

You have made no comment or acknowledged state partitioning, either

**Comment by ghostwords** [comment]
Privacy Badger blocking efficacy:

>Notably this means that the combination of Privacy Badger and uBlock Origin block more than each extension on their own, highlighting again that they are indeed complimentary.

-- [Understanding the Performance Costs and Benefits of Privacy-focused Browser Extensions](https://kevin.borgolte.me/files/pdf/www2020-privacy-extensions.pdf)

----

State partioning and the need to block trackers:

>In a series of posts on this blog in 2017/18, we revealed how web trackers exfiltrate personal information from web pages, browser password managers, form inputs, and the Facebook Login API.

-- https://freedom-to-tinker.com/tag/noboundaries/

>Partitioning isn’t a panacea. There are many kinds of privacy harm that partitioning doesn’t protect users against (e.g., browser fingerprinting, cookie syncing from user identifiers, IP based tracking, etc.).

-- https://brave.com/privacy-updates/13-pool-party-side-channels/

**Comment by Thorin-Oakenpants** [comment]
Not a single instance of the word "partition" in the linked PDF
- FWIW I got around to testing and edited the arkenfox description for PB - it's still redundant for Firefox users with dFPI

As for pool parties. feel free to read the bugzilla - https://bugzilla.mozilla.org/show_bug.cgi?id=1730797#c18 . No one said state partitioning encompasses absolutely everything, but clearly, partitioning in this case can be used to help mitigate

**Comment by ghostwords** [comment]
Do you see how "[effective privacy blocker extension] is redundant [...] with dFPI" and "No one said state partitioning encompasses absolutely everything" are contradictory statements?

**Comment by Thorin-Oakenpants** [comment]
PDF
> Notably this means that the combination of Privacy Badger and uBlock Origin block more than each extension on their own, highlighting again that they are indeed complimentary.

The paper is from early 2020. The PB version is `2019.7.1`. Heuristics were enabled by default. There is no mention of state partitioning (which wasn't default in any browsers excluding Tor Browser). No one is arguing that PB's heuristics can't learn and pick up new items - because blocklists are just a form of enumerating badness.

The argument was that state partitioning makes it redundant 

> Do you see ... are contradictory statements?

They're not contradictory if you take them in context of the state (pun not intended) of tracking today and what partitioning in Firefox can cover today (e.g. turn on dFPI). Did you read the bugzilla comment?

Seriously, what use is PB in non-heuristic mode (and you shouldn't use heuristic mode, you disabled it for a reason, even wrote a blog post warning everyone) given it's tiny blocklist - I wouldn't be surprised if these aren't already constantly checked and merged into other blocklists

**Comment by ghostwords** [comment]
State partitioning doesn't protect against:
- IP tracking
- browser fingerprinting
- querystring ID syncing
- personal information exfiltration
...
- slow websites
- wasting CPU
- wasting bandwidth

Blocking trackers does.

**Comment by ghostwords** [comment]
Privacy Badger is an effective tracker blocker with or without enabling local learning. You should absolutely enable local learning if you are comfortable with the tradeoffs. We wrote the aforementioned EFF blog post to explain the tradeoffs.

**Comment by Thorin-Oakenpants** [comment]
IDK why you are making strawman arguments. State partitioning is to deal with state tracking, so why bring up other tracking methods such as redirects, query params, or STATELESS tracking. No one has claimed state partitioning is a cure-all for all tracking or linkability issues.

**Comment by ghostwords** [comment]
>Most (not all) cross-domain state tracking has been neutralized (and arkenfox sanitizes per session). In that context, Privacy Badger is obsolete, and even uBlock Origin is almost a luxury

**Comment by Thorin-Oakenpants** [comment]
Correct. uBlock Origin as a tracker blocker is obsolete. But it is a content blocker (blocks ads which is attack surface, strips url params, etc). What is your point? That I didn't make uBO also redundant?

**Comment by ghostwords** [comment]
What is stateless tracking if not tracking?

**Comment by Thorin-Oakenpants** [comment]
State and stateless tracking are subsets of "tracking". State tracking has a solution called state partitioning. Stateless tracking has several solutions - the first being blocklists (enumerating badness), but the ultimate solution is making fingerprinting useless (such as Brave's shield, or Tor Browser's RFP), but even those need to deal with the network fingerprint (IP: at least Tor Browser has Tor). There are also other methods of tracking/linkability, such as url params - for each of which there are other solutions (with different degrees of success)

Why are you trying to make "state partitioning" look bad because it doesn't cover non-state issues?

Again, what does PB in non-heuristic mode bring to the party? Does it strip url tracking params? Does it hide your IP? Does it solve the pool party problem? Does it defeat all fingerprinting? The answer to that last one is it can't - blocking only goes so far and is trivial to bypass.

Look, I get that you have invested interest in PB. Just agree to disagree rather than making more strawmen

**Comment by ghostwords** [comment]
I was trying to understand where you're coming from so I can consider your perspective.

What does Privacy Badger bring to the table? Why, that brings us all the way back to https://github.com/tycrek/degoogle/issues/426#issuecomment-983737389

**Comment by ghostwords** [comment]
I truly don't understand how you get to the following when blocking trackers stops tracking, both stateful and stateless.

>uBlock Origin as a tracker blocker is obsolete

**Comment by ghostwords** [comment]
Why wouldn't you welcome a layered approach where dFPI is good and an effective tracker blocker is also good?

**Comment by Thorin-Oakenpants** [comment]
edit: apologies if this is a bit long. FP short of fingerprint. I've stuck it behind a click

> when blocking trackers stops tracking, both stateful and stateless

<details><summary>click me for long reply</summary><p>

First, when I talk about FPing, I'm talking about non-IP metrics. I leave the IP part to others and Tor/I2P/etc. But it _is_ an integral part of a FP, just not my words below

uBO blocks a lot of these stateless scripts. ETP has a fingerprintERs blocklist (from openWMP). It's enumerating badness and not a final solution. Don't get me wrong, a final solution is hard (but doable: enough to make FPing extremely hard if not outright useless and educating users a little because some breakage/inconvenience will happen - when you break web standards this is what you will sometimes get). Blocking absolutely plays a part (for now)

On fingerprinting. If a user does nothing, they are unique. There are essentially two types of scripts: naive and advanced. Naive scripts swallow a poison pill (a randomized value) and an advanced script doesn't (all randomizing can be detected, btw). Advanced scripts can be simple or extremely complex, but we can ignore that for now. The more you randomize the more the chance of fooling a script, but the tail is long and thin. Brave randomizes lots, RFP only one (that matters).

When fighting FPing, there are numerous steps. The first is block known scripts that can be blocked. But ultimately (for all scripts) you need to actually address the FPing of metrics (passive and active), one by one, until they are rendered to equivalency of something bigger (such as OS) or the same for everyone (randomizing is detected and returned as a static value) - always in a set (like "TB linux users"), there are some things you cannot hide. There are numerous methods to do this. Ultimately advanced scripts require a crowd to hide in, naive scripts don't. Only Tor Browser has covered enough metrics to make FPing extremely hard (but without a real world study it's hard to know for sure), let alone address the IP issue.

The best non Tor Browser browsers can do is block known scripts and fool naive ones that get through or can't be blocked. An advanced script will find you unique in my experience - except RFP comes very close: they just need webGL covered and a few tweaks, but RFP is not front facing and Firefox does not have a "crowd" by default - this is all being worked on, and has been for the last five years (and I've been involved with that) to bring it to Firefox users - imagine 220mn potential users as a crowd, compared with TB's crowd of several millions

Now linking FPs is another science in it's own. FPs mutate over time. Most scripts were/are naive, but that is changing. Scripts are now using JS attack templates - such as TZP/CreepJS pulling large arrays of properties and more. And a FP is only a snapshot of data at a given point - and items can be included/excluded at will such as knowing the zoom is not the same, you can reverse values such as screen, or ignore others that are not linear such as textmetrics.

That said, universality of a script is required. Different canvas tests can't be linked for example. And most known scripts that have any universality worth a dose of salt are known and blocked (AFAIK). But it's a dynamic world out there. The more we fight fingerprinting, the more they try harder - but it is very telling that a lot of them have openly claimed and stated that Tor Browser is too hard and given up

So keeping that all in mind, when talking about the threat of fingerprinting in tracking/linkability in Firefox with arkenfox (which uses RFP or encourages users who do not want RFP to use CanvasBlocker = naive scripts), and thinking about how (most?) known FPing scripts are naive and/or are already blocked

So coming back to "Blocking absolutely plays a part (for now)", it absolutely cannot hurt and most likely helps (not all scripts are naive). But defeating FPing is going to take a lot more than tracker blocking - in the meantime RFP has a bigger impact IMO (all naive scripts) and ETP/uBO already have blocklists. _**Does PB detect new fingerprinting scripts when local learning is disabled?**_ If not then I consider it redundant (the blocklist is probably already covered in uBO/ETP?) and the threat is over-hyped (universality and all naive scripts already covered) - and that doesn't even touch on the fact that your IP is not covered (even if you use a VPN and hop-a-lot).

So it's all a bit moot IMO. Stateless tracking is not the threat most people seem it is (not yet) and most of it is already mitigated but it's _almost_ a lost cause anyway as it still doesn't address IPs

tl;dr: I don't care about FPing except a final solution, and users of Firefox etc shouldn't either - if their threat model calls for it, they should use Tor Browser

</p></details>

There's a lot more to FPing of course, and I'm not here to lecture you on it - just giving you my (generalized) insight. Again, apologies for the length

---

Oh, I do welcome a layered approach, but I fail to see how the (tiny) blocklist in PB adds any value in non-heuristic mode.

Looking at https://github.com/tycrek/degoogle/issues/426#issuecomment-983737389 
> regardless of whether you enable learning locally, Privacy Badger can automatically discover new trackers that list-based blockers don’t know about

So the "enable local learning" checkbox does not disable local learning? Can you enlighten me (it's not like I monitor loads of extensions)

> Privacy Badger comes with privacy features like click-to-activate replacements for potentially useful trackers (video players, comments sections like Disqus, etc.), and link cleaning on Facebook and Google.

I would have to compare with the Smart Blocking shims and heuristics for cross-domain logins in dFPI (from memory they already covers the big players like the three you mentioned) - but again, I don't monitor or test lots of extensions. Arkenfox already disables sets a lot of prefs, including autoplay - is this the same as click-to-activate for video players?

> Privacy Badger is also a political tool. 

Agreed. And I love EFF. But it has zero bearing on my decisions

> Privacy Badger sends the Global Privacy Control signal to opt you out of data sharing and selling, and the Do Not Track signal to tell companies not to track you. If trackers ignore your wishes, Privacy Badger will learn to block them

GPC will get enabled in time - as it is right now, it makes users FP stand out on firefox (not knocking the effort and users are already unique - so a bit of a moot point - see my long ramble above). DNT is sent regardless with ETP's Tracking Protection (which is part of ETP strict). I know that doesn't deal with the "if trackers ignore your wishes" - in terms of DNT that's going to be almost every site out there. But this is moot (IMO) with state partitioning

So color me intrigued that twice in that post you said that PB learns, once even stating that it does that with local learning off.

I get there are different mechanisms you can use: e.g. if a tracker does respect DNT it can be auto-added vs injecting FP code into DOM, but can you enlighten me.

**Comment by ghostwords** [comment]
I appreciate the long response but it's not easy to understand where you actually stand. From what I can tell you don't actually think that "dFPI makes tracker blockers obsolete".

I'll sign off here with two things I said earlier in https://github.com/tycrek/degoogle/issues/426#issuecomment-984636914.

>uBlock Origin has its own mission and feature set that partially overlaps with but is distinct from Privacy Badger's mission and feature set.
>
>Why push for a monoculture where all privacy tools are based on manually assembled lists?

**Comment by Thorin-Oakenpants** [comment]
> it's not easy to understand where you actually stand

I'll wrap it up with this. Probably a better way of saying it is that the ROI (in non-heuristic mode) is going to be minimal (if that) to zero (with some browser pref changes). I was talking about state tracking and did not _fully_ consider stateless - but without a stateless heuristic, I'll stand by that. A static list isn't going to add anything meaningful, and stateless is "overrated" as a threat (for now) and is best handled by already existing blocklists and fooling naive scripts (with better solutions coming)

I wouldn't say it's a monoculture - it's built-in browser solutions with as few extensions as required (e.g. uBO especially in a medium+ mode)

**Comment by tycrek** [comment]
Apologies for leaving this open for so long. What is the state of Privacy Badger in 2023?

**Comment by freddy-m** [comment]
@tycrek its basically not needed. Almost all the browser extensions listed, however helpful they may be, in combination just end up making you less private. I personally only use uBlock Origin and a password manager and thats it. Less is more, in these cases.

**Comment by ghostwords** [comment]
It's still what I wrote above in https://github.com/tycrek/degoogle/issues/426#issuecomment-983737389, plus a bunch of improvements such as [improved link tracking protection on Google sites](https://www.eff.org/deeplinks/2023/09/new-privacy-badger-prevents-google-mangling-more-your-links-and-invading-your).

Yes, it doesn't make a lot of sense to use two or more list-based ad blockers together. Just use the best one. Privacy Badger is not list-based and is not an ad blocker.

Privacy Badger is good by itself and is also good with uBlock Origin. uBlock Origin is good by itself and is also good with Privacy Badger.

**Comment by tycrek** [comment]
> in combination just end up making you less private.

I won't pretend to fully understand the state conversation above; but how exactly does using two make you less private than just one? If either one is modifying the page, surely that's just as detectable as two, no?

The conversation above also talks about Arkenfox. Does uBlock+Arkenfox cover the things that PB does?

If uBO and PB had identical featuresets for the same purpose (such as the case of HTTPS Everywhere and HTTPZ), I'd be more inclined to remove it, It may also be worth adding some of the links shared here in that section.

**Comment by ghostwords** [comment]
The big idea is that each extension that modifies pages can be detected separately by a page. So if you have a collection of five such extensions, you stand out more than somebody who has just one popular extension. This basically falls under browser fingerprinting.

The thing though is that Privacy Badger (and probably some ad blocker lists) specifically aims to block browser fingerprinters. It's important to maintain perspective. I think it absolutely makes sense to block more trackers from following you across the Web at the expense of hypothetically making yourself stand out more to websites.

**Comment by freddy-m** [comment]
> If uBO and PB had identical featuresets for the same purpose (such as the case of HTTPS Everywhere and HTTPZ), I'd be more inclined to remove it. It may also be worth adding some of the links shared here in that section.

This seems like a fairly decent compromise. I'd still ideally want to remove most of the other browser extensions, they are increasingly obsolete. I'd finish by adding that there is a reason why the only extension listed on Privacy Guides is uBlock Origin :)
